File: app/src/main/java/dev/ragnarok/fenrir/media/voice/ExoVoicePlayerSensored.java
package dev.ragnarok.fenrir.media.voice;

import static dev.ragnarok.fenrir.util.Objects.isNull;
import static dev.ragnarok.fenrir.util.Objects.nonNull;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.os.Build;
import android.os.PowerManager;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.exoplayer2.C;
import com.google.android.exoplayer2.DefaultRenderersFactory;
import com.google.android.exoplayer2.ExoPlayer;
import com.google.android.exoplayer2.PlaybackException;
import com.google.android.exoplayer2.Player;
import com.google.android.exoplayer2.audio.AudioAttributes;
import com.google.android.exoplayer2.source.MediaSource;
import com.google.android.exoplayer2.source.ProgressiveMediaSource;

import dev.ragnarok.fenrir.AccountType;
import dev.ragnarok.fenrir.Constants;
import dev.ragnarok.fenrir.media.exo.ExoUtil;
import dev.ragnarok.fenrir.model.ProxyConfig;
import dev.ragnarok.fenrir.model.VoiceMessage;
import dev.ragnarok.fenrir.player.MusicPlaybackController;
import dev.ragnarok.fenrir.settings.Settings;
import dev.ragnarok.fenrir.util.Logger;
import dev.ragnarok.fenrir.util.Optional;
import dev.ragnarok.fenrir.util.Utils;

public class ExoVoicePlayerSensored implements IVoicePlayer, SensorEventListener {

    private final Context app;
    private final ProxyConfig proxyConfig;
    private final SensorManager sensorManager;
    private final Sensor proxym;
    private final PowerManager.WakeLock proximityWakelock;
    private final MusicIntentReceiver headset;
    private ExoPlayer exoPlayer;
    private int status;
    private AudioEntry playingEntry;
    private boolean supposedToBePlaying;
    private IPlayerStatusListener statusListener;
    private IErrorListener errorListener;
    private boolean isProximityNear;
    private boolean isPlaying;
    private boolean HasPlaying;
    private boolean playbackSpeed;
    private boolean Registered;
    private boolean ProximitRegistered;
    private boolean isHeadset;

    public ExoVoicePlayerSensored(Context context, ProxyConfig config) {
        app = context.getApplicationContext();
        proxyConfig = config;
        status = STATUS_NO_PLAYBACK;
        headset = new MusicIntentReceiver();

        sensorManager = (SensorManager) app.getSystemService(Context.SENSOR_SERVICE);
        proxym = sensorManager.getDefaultSensor(Sensor.TYPE_PROXIMITY);
        proximityWakelock = ((PowerManager) app.getSystemService(Context.POWER_SERVICE)).newWakeLock(PowerManager.PROXIMITY_SCREEN_OFF_WAKE_LOCK, "fenrir:voip=proxim");
        Registered = false;
        ProximitRegistered = false;
        HasPlaying = false;
    }

    private static boolean isOpusSupported() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q || Settings.get().other().isEnable_native() || Settings.get().other().getFFmpegPlugin() != 0;
    }

    @Override
    public boolean toggle(int id, VoiceMessage audio) {
        if (nonNull(playingEntry) && playingEntry.getId() == id) {
            setSupposedToBePlaying(!isSupposedToPlay());
            return false;
        }

        release();

        playingEntry = new AudioEntry(id, audio);
        supposedToBePlaying = true;

        preparePlayer();
        return true;
    }

    private void RegisterCallBack() {
        if (Registered)
            return;
        try {
            Registered = true;
            if (MusicPlaybackController.isPlaying() || MusicPlaybackController.isPreparing()) {
                MusicPlaybackController.notifyForegroundStateChanged(app, true);
                MusicPlaybackController.playOrPause();
                HasPlaying = true;
            }
            isProximityNear = false;
            isHeadset = false;
            exoPlayer.setAudioAttributes(new AudioAttributes.Builder().setContentType(C.CONTENT_TYPE_MUSIC).setUsage(C.USAGE_MEDIA).build(), true);
            sensorManager.registerListener(this, proxym, SensorManager.SENSOR_DELAY_NORMAL);
            IntentFilter filter = new IntentFilter(Intent.ACTION_HEADSET_PLUG);
            app.registerReceiver(headset, filter);
        } catch (Exception ignored) {
        }
    }

    private void UnRegisterCallBack() {
        if (!Registered)
            return;
        try {
            Registered = false;
            sensorManager.unregisterListener(this);
            app.unregisterReceiver(headset);
            if (HasPlaying) {
                MusicPlaybackController.playOrPause();
                MusicPlaybackController.notifyForegroundStateChanged(app, false);
            }
            HasPlaying = false;
            if (ProximitRegistered) {
                ProximitRegistered = false;
                proximityWakelock.release();
            }
            isProximityNear = false;
            isHeadset = false;
            isPlaying = false;
        } catch (Exception ignored) {
        }
    }

    private void setStatus(int status) {
        if (this.status != status) {
            this.status = status;

            if (nonNull(statusListener)) {
                statusListener.onPlayerStatusChange(status);
            }
        }
    }

    private void preparePlayer() {
        isProximityNear = false;
        isHeadset = false;
        isPlaying = false;
        setStatus(STATUS_PREPARING);

        int extensionRenderer = DefaultRenderersFactory.EXTENSION_RENDERER_MODE_OFF;
        switch (Settings.get().other().getFFmpegPlugin()) {
            case 0:
                extensionRenderer = DefaultRenderersFactory.EXTENSION_RENDERER_MODE_OFF;
                break;
            case 1:
                extensionRenderer = DefaultRenderersFactory.EXTENSION_RENDERER_MODE_ON;
                break;
            case 2:
                extensionRenderer = DefaultRenderersFactory.EXTENSION_RENDERER_MODE_PREFER;
                break;
        }
        exoPlayer = new ExoPlayer.Builder(app, new DefaultRenderersFactory(app).setExtensionRendererMode(extensionRenderer)).build();
        exoPlayer.setWakeMode(C.WAKE_MODE_NETWORK);

        // DefaultBandwidthMeter bandwidthMeterA = new DefaultBandwidthMeter();
        // Produces DataSource instances through which media data is loaded.
        // DataSource.Factory dataSourceFactory = new DefaultDataSourceFactory(this, Util.getUserAgent(this, "exoplayer2example"), bandwidthMeterA);
        // DefaultDataSourceFactory dataSourceFactory = new DefaultDataSourceFactory(App.getInstance(), Util.getUserAgent(App.getInstance(), "exoplayer2example"), bandwidthMeterA);

        String userAgent = Constants.USER_AGENT(AccountType.BY_TYPE);

        // This is the MediaSource representing the media to be played:
        // FOR SD CARD SOURCE:
        // MediaSource videoSource = new ExtractorMediaSource(mp4VideoUri, dataSourceFactory, extractorsFactory, null, null);
        // FOR LIVESTREAM LINK:

        String url = isOpusSupported() ? Utils.firstNonEmptyString(playingEntry.getAudio().getLinkOgg(), playingEntry.getAudio().getLinkMp3()) : playingEntry.getAudio().getLinkMp3();

        MediaSource mediaSource = new ProgressiveMediaSource.Factory(Utils.getExoPlayerFactory(userAgent, proxyConfig)).createMediaSource(Utils.makeMediaItem(url));
        exoPlayer.setRepeatMode(Player.REPEAT_MODE_OFF);
        exoPlayer.addListener(new Player.Listener() {
            @Override
            public void onPlaybackStateChanged(@Player.State int state) {
                onInternalPlayerStateChanged(state);
            }

            @Override
            public void onPlayWhenReadyChanged(boolean playWhenReady, @Player.PlayWhenReadyChangeReason int reason) {
                if (isPlaying != playWhenReady) {
                    isPlaying = playWhenReady;
                    if (isPlaying) {
                        RegisterCallBack();
                    } else {
                        UnRegisterCallBack();
                    }
                }
            }

            @Override
            public void onPlayerError(@NonNull PlaybackException error) {
                onExoPlayerException(error);
                UnRegisterCallBack();
            }
        });

        exoPlayer.setPlayWhenReady(supposedToBePlaying);
        exoPlayer.setMediaSource(mediaSource);
        exoPlayer.setPlaybackSpeed(playbackSpeed ? 2f : 1f);
        exoPlayer.prepare();
    }

    private void onExoPlayerException(@NonNull PlaybackException e) {
        if (nonNull(errorListener)) {
            errorListener.onPlayError(new PrepareException(e));
        }
    }

    private void onInternalPlayerStateChanged(@Player.State int state) {
        Logger.d("ExoVoicePlayer", "onInternalPlayerStateChanged, state: " + state);

        switch (state) {
            case Player.STATE_READY:
                setStatus(STATUS_PREPARED);
                break;
            case Player.STATE_ENDED:
                setSupposedToBePlaying(false);
                exoPlayer.seekTo(0);
                UnRegisterCallBack();
                break;
            case Player.STATE_BUFFERING:
            case Player.STATE_IDLE:
                break;
        }
    }

    private void setSupposedToBePlaying(boolean supposedToBePlaying) {
        this.supposedToBePlaying = supposedToBePlaying;

        if (supposedToBePlaying) {
            ExoUtil.startPlayer(exoPlayer);
        } else {
            ExoUtil.pausePlayer(exoPlayer);
        }
    }

    private long getDuration() {
        if (isNull(playingEntry) || isNull(playingEntry.getAudio()) || playingEntry.getAudio().getDuration() == 0) {
            return Math.max(exoPlayer.getDuration(), 1L);
        }
        return playingEntry.getAudio().getDuration() * 1000L;
    }

    @Override
    public float getProgress() {
        if (isNull(exoPlayer)) {
            return 0f;
        }

        if (status != STATUS_PREPARED) {
            return 0f;
        }

        long duration = getDuration();
        long position = exoPlayer.getCurrentPosition();
        return (float) position / (float) duration;
    }

    @Override
    public void setCallback(@Nullable IPlayerStatusListener listener) {
        statusListener = listener;
    }

    @Override
    public void setErrorListener(@Nullable IErrorListener errorListener) {
        this.errorListener = errorListener;
    }

    @Override
    public Optional<Integer> getPlayingVoiceId() {
        return isNull(playingEntry) ? Optional.empty() : Optional.wrap(playingEntry.getId());
    }

    @Override
    public boolean isSupposedToPlay() {
        return supposedToBePlaying;
    }

    @Override
    public boolean isPlaybackSpeed() {
        return playbackSpeed;
    }

    @Override
    public void togglePlaybackSpeed() {
        if (nonNull(exoPlayer)) {
            playbackSpeed = !playbackSpeed;
            exoPlayer.setPlaybackSpeed(playbackSpeed ? 2f : 1f);
        }
    }

    @Override
    public void release() {
        try {
            if (nonNull(exoPlayer)) {
                exoPlayer.stop();
                exoPlayer.release();
                UnRegisterCallBack();
            }
        } catch (Exception ignored) {
        }
    }

    @Override
    public void onSensorChanged(SensorEvent event) {
        if (isHeadset)
            return;
        if (event.sensor.getType() == Sensor.TYPE_PROXIMITY) {
            boolean newIsNear = event.values[0] < Math.min(event.sensor.getMaximumRange(), 3);
            if (newIsNear != isProximityNear) {
                isProximityNear = newIsNear;
                try {
                    if (isProximityNear) {
                        exoPlayer.setAudioAttributes(new AudioAttributes.Builder().setContentType(C.CONTENT_TYPE_SPEECH).setUsage(C.USAGE_VOICE_COMMUNICATION).build(), false);
                        if (!ProximitRegistered) {
                            ProximitRegistered = true;
                            proximityWakelock.acquire(10 * 60 * 1000L /*10 minutes*/);
                        }
                    } else {
                        if (ProximitRegistered) {
                            ProximitRegistered = false;
                            proximityWakelock.release(1); // this is non-public API before L
                        }
                        exoPlayer.setAudioAttributes(new AudioAttributes.Builder().setContentType(C.CONTENT_TYPE_MUSIC).setUsage(C.USAGE_MEDIA).build(), true);
                    }
                } catch (Exception ignored) {
                }
            }
        }
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {
    }

    private class MusicIntentReceiver extends BroadcastReceiver {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (java.util.Objects.equals(intent.getAction(), Intent.ACTION_HEADSET_PLUG)) {
                int state = intent.getIntExtra("state", -1);
                switch (state) {
                    case 0:
                        if (isHeadset) {
                            isHeadset = false;
                        }
                        break;
                    case 1:
                        if (!isHeadset) {
                            isHeadset = true;
                            isProximityNear = false;
                            try {
                                if (ProximitRegistered) {
                                    ProximitRegistered = false;
                                    proximityWakelock.release(1); // this is non-public API before L
                                }
                                exoPlayer.setAudioAttributes(new AudioAttributes.Builder().setContentType(C.CONTENT_TYPE_MUSIC).setUsage(C.USAGE_MEDIA).build(), true);
                            } catch (Exception ignored) {
                            }
                        }
                        break;
                    default:
                        break;
                }
            }
        }
    }
}

File: app/src/main/java/dev/ragnarok/fenrir/media/voice/ExoVoicePlayerSensored.kt
package dev.ragnarok.fenrir.media.voice

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Build
import android.os.PowerManager
import com.google.android.exoplayer2.*
import com.google.android.exoplayer2.Player.PlayWhenReadyChangeReason
import com.google.android.exoplayer2.audio.AudioAttributes
import com.google.android.exoplayer2.source.MediaSource
import com.google.android.exoplayer2.source.ProgressiveMediaSource
import dev.ragnarok.fenrir.AccountType
import dev.ragnarok.fenrir.Constants.USER_AGENT
import dev.ragnarok.fenrir.media.exo.ExoUtil.pausePlayer
import dev.ragnarok.fenrir.media.exo.ExoUtil.startPlayer
import dev.ragnarok.fenrir.media.music.MusicPlaybackController
import dev.ragnarok.fenrir.media.music.MusicPlaybackController.isPreparing
import dev.ragnarok.fenrir.media.music.MusicPlaybackController.notifyForegroundStateChanged
import dev.ragnarok.fenrir.media.music.MusicPlaybackController.playOrPause
import dev.ragnarok.fenrir.media.voice.IVoicePlayer.IErrorListener
import dev.ragnarok.fenrir.media.voice.IVoicePlayer.IPlayerStatusListener
import dev.ragnarok.fenrir.model.ProxyConfig
import dev.ragnarok.fenrir.model.VoiceMessage
import dev.ragnarok.fenrir.settings.Settings
import dev.ragnarok.fenrir.util.Logger.d
import dev.ragnarok.fenrir.util.Optional
import dev.ragnarok.fenrir.util.Optional.Companion.empty
import dev.ragnarok.fenrir.util.Optional.Companion.wrap
import dev.ragnarok.fenrir.util.Utils.firstNonEmptyString
import dev.ragnarok.fenrir.util.Utils.getExoPlayerFactory
import dev.ragnarok.fenrir.util.Utils.makeMediaItem

class ExoVoicePlayerSensored(context: Context, config: ProxyConfig?) : IVoicePlayer,
    SensorEventListener {
    private val app: Context = context.applicationContext
    private val proxyConfig: ProxyConfig? = config
    private val sensorManager: SensorManager?
    private val proxym: Sensor?
    private val proximityWakelock: PowerManager.WakeLock?
    private val headset: MusicIntentReceiver
    private var exoPlayer: ExoPlayer? = null
    private var status: Int
    private var playingEntry: AudioEntry? = null
    private var supposedToBePlaying = false
    private var statusListener: IPlayerStatusListener? = null
    private var errorListener: IErrorListener? = null
    private var isProximityNear = false
    private var isPlaying = false
    private var HasPlaying: Boolean
    private var Registered: Boolean
    private var ProximitRegistered: Boolean
    private var isHeadset = false
    private var playbackSpeed = false
    override fun toggle(id: Int, audio: VoiceMessage): Boolean {
        if (playingEntry?.id == id) {
            setSupposedToBePlaying(!isSupposedToPlay)
            return false
        }
        release()
        playingEntry = AudioEntry(id, audio)
        supposedToBePlaying = true
        preparePlayer()
        return true
    }

    private fun RegisterCallBack() {
        if (Registered) return
        try {
            Registered = true
            if (MusicPlaybackController.isPlaying || isPreparing) {
                notifyForegroundStateChanged(app, true)
                playOrPause()
                HasPlaying = true
            }
            isProximityNear = false
            isHeadset = false
            exoPlayer?.setAudioAttributes(
                AudioAttributes.Builder().setContentType(C.CONTENT_TYPE_MUSIC).setUsage(
                    C.USAGE_MEDIA
                ).build(), true
            )
            sensorManager?.registerListener(this, proxym, SensorManager.SENSOR_DELAY_NORMAL)
            val filter = IntentFilter(Intent.ACTION_HEADSET_PLUG)
            app.registerReceiver(headset, filter)
        } catch (ignored: Exception) {
        }
    }

    private fun UnRegisterCallBack() {
        if (!Registered) return
        try {
            Registered = false
            sensorManager?.unregisterListener(this)
            app.unregisterReceiver(headset)
            if (HasPlaying) {
                playOrPause()
                notifyForegroundStateChanged(app, false)
            }
            HasPlaying = false
            if (ProximitRegistered) {
                ProximitRegistered = false
                proximityWakelock?.release()
            }
            isProximityNear = false
            isHeadset = false
            isPlaying = false
        } catch (ignored: Exception) {
        }
    }

    private fun setStatus(status: Int) {
        if (this.status != status) {
            this.status = status
            statusListener?.onPlayerStatusChange(status)
        }
    }

    private fun preparePlayer() {
        isProximityNear = false
        isHeadset = false
        isPlaying = false
        setStatus(IVoicePlayer.STATUS_PREPARING)
        var extensionRenderer = DefaultRenderersFactory.EXTENSION_RENDERER_MODE_OFF
        when (Settings.get().other().fFmpegPlugin) {
            0 -> extensionRenderer = DefaultRenderersFactory.EXTENSION_RENDERER_MODE_OFF
            1 -> extensionRenderer = DefaultRenderersFactory.EXTENSION_RENDERER_MODE_ON
            2 -> extensionRenderer = DefaultRenderersFactory.EXTENSION_RENDERER_MODE_PREFER
        }
        exoPlayer = ExoPlayer.Builder(
            app,
            DefaultRenderersFactory(app).setExtensionRendererMode(extensionRenderer)
        ).build()
        exoPlayer?.setWakeMode(C.WAKE_MODE_NETWORK)
        val userAgent = USER_AGENT(AccountType.BY_TYPE)
        val url = if (isOpusSupported) firstNonEmptyString(
            playingEntry?.audio?.linkOgg, playingEntry?.audio?.linkMp3
        ) else playingEntry?.audio?.linkMp3
        val mediaSource: MediaSource =
            ProgressiveMediaSource.Factory(getExoPlayerFactory(userAgent, proxyConfig))
                .createMediaSource(
                    makeMediaItem(url)
                )
        exoPlayer?.repeatMode = Player.REPEAT_MODE_OFF
        exoPlayer?.addListener(object : Player.Listener {
            override fun onPlaybackStateChanged(state: @Player.State Int) {
                onInternalPlayerStateChanged(state)
            }

            override fun onPlayWhenReadyChanged(
                playWhenReady: Boolean,
                reason: @PlayWhenReadyChangeReason Int
            ) {
                if (isPlaying != playWhenReady) {
                    isPlaying = playWhenReady
                    if (isPlaying) {
                        RegisterCallBack()
                    } else {
                        UnRegisterCallBack()
                    }
                }
            }

            override fun onPlayerError(error: PlaybackException) {
                onExoPlayerException(error)
                UnRegisterCallBack()
            }
        })
        exoPlayer?.playWhenReady = supposedToBePlaying
        exoPlayer?.setMediaSource(mediaSource)
        exoPlayer?.setPlaybackSpeed(if (playbackSpeed) 2f else 1f)
        exoPlayer?.prepare()
    }

    private fun onExoPlayerException(e: PlaybackException) {
        errorListener?.onPlayError(e)
    }

    private fun onInternalPlayerStateChanged(state: @Player.State Int) {
        d("ExoVoicePlayer", "onInternalPlayerStateChanged, state: $state")
        when (state) {
            Player.STATE_READY -> setStatus(IVoicePlayer.STATUS_PREPARED)
            Player.STATE_ENDED -> {
                setSupposedToBePlaying(false)
                exoPlayer?.seekTo(0)
                UnRegisterCallBack()
            }
            Player.STATE_BUFFERING, Player.STATE_IDLE -> {}
        }
    }

    private fun setSupposedToBePlaying(supposedToBePlaying: Boolean) {
        this.supposedToBePlaying = supposedToBePlaying
        if (supposedToBePlaying) {
            startPlayer(exoPlayer)
        } else {
            pausePlayer(exoPlayer)
        }
    }

    private val duration: Long
        get() = if (playingEntry == null || playingEntry?.audio == null || playingEntry?.audio?.duration == 0) {
            exoPlayer?.duration?.coerceAtLeast(1L) ?: 1
        } else (playingEntry?.audio?.duration?.times(1000L)) ?: 1

    override val progress: Float
        get() = when {
            exoPlayer == null -> {
                0f
            }
            status != IVoicePlayer.STATUS_PREPARED -> {
                0f
            }
            else -> {
                val duration = duration
                val position = exoPlayer?.currentPosition ?: 0
                position.toFloat() / duration.toFloat()
            }
        }


    override fun setCallback(listener: IPlayerStatusListener?) {
        statusListener = listener
    }

    override fun setErrorListener(errorListener: IErrorListener?) {
        this.errorListener = errorListener
    }

    override val playingVoiceId: Optional<Int>
        get() = playingEntry?.let { wrap(it.id) } ?: empty()


    override val isSupposedToPlay: Boolean
        get() = supposedToBePlaying

    override val isPlaybackSpeed: Boolean
        get() = playbackSpeed

    override fun togglePlaybackSpeed() {
        if (exoPlayer != null) {
            playbackSpeed = !playbackSpeed
            exoPlayer?.setPlaybackSpeed(if (playbackSpeed) 2f else 1f)
        }
    }

    override fun release() {
        try {
            if (exoPlayer != null) {
                exoPlayer?.stop()
                exoPlayer?.release()
                UnRegisterCallBack()
            }
        } catch (ignored: Exception) {
        }
    }

    override fun onSensorChanged(event: SensorEvent) {
        if (isHeadset) return
        if (event.sensor.type == Sensor.TYPE_PROXIMITY) {
            val newIsNear = event.values[0] < event.sensor.maximumRange.coerceAtMost(3f)
            if (newIsNear != isProximityNear) {
                isProximityNear = newIsNear
                try {
                    if (isProximityNear) {
                        exoPlayer?.setAudioAttributes(
                            AudioAttributes.Builder().setContentType(C.CONTENT_TYPE_SPEECH)
                                .setUsage(
                                    C.USAGE_VOICE_COMMUNICATION
                                ).build(), false
                        )
                        if (!ProximitRegistered) {
                            ProximitRegistered = true
                            proximityWakelock?.acquire(10 * 60 * 1000L /*10 minutes*/)
                        }
                    } else {
                        if (ProximitRegistered) {
                            ProximitRegistered = false
                            proximityWakelock?.release(1) // this is non-public API before L
                        }
                        exoPlayer?.setAudioAttributes(
                            AudioAttributes.Builder().setContentType(C.CONTENT_TYPE_MUSIC).setUsage(
                                C.USAGE_MEDIA
                            ).build(), true
                        )
                    }
                } catch (ignored: Exception) {
                }
            }
        }
    }

    override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {}
    private inner class MusicIntentReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if (intent.action == Intent.ACTION_HEADSET_PLUG) {
                when (intent.getIntExtra("state", -1)) {
                    0 -> if (isHeadset) {
                        isHeadset = false
                    }
                    1 -> if (!isHeadset) {
                        isHeadset = true
                        isProximityNear = false
                        try {
                            if (ProximitRegistered) {
                                ProximitRegistered = false
                                proximityWakelock?.release(1) // this is non-public API before L
                            }
                            exoPlayer?.setAudioAttributes(
                                AudioAttributes.Builder().setContentType(
                                    C.CONTENT_TYPE_MUSIC
                                ).setUsage(C.USAGE_MEDIA).build(), true
                            )
                        } catch (ignored: Exception) {
                        }
                    }
                    else -> {}
                }
            }
        }
    }

    companion object {
        private val isOpusSupported: Boolean
            get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q || Settings.get()
                .other().isEnable_native || Settings.get().other().fFmpegPlugin != 0
    }

    init {
        status = IVoicePlayer.STATUS_NO_PLAYBACK
        headset = MusicIntentReceiver()
        sensorManager = app.getSystemService(Context.SENSOR_SERVICE) as SensorManager?
        proxym = sensorManager?.getDefaultSensor(Sensor.TYPE_PROXIMITY)
        proximityWakelock =
            (app.getSystemService(Context.POWER_SERVICE) as PowerManager?)?.newWakeLock(
                PowerManager.PROXIMITY_SCREEN_OFF_WAKE_LOCK,
                "fenrir:voip=proxim"
            )
        Registered = false
        ProximitRegistered = false
        HasPlaying = false
    }
}
